opendatahub-io/odh-dashboard
# logic for deprecated images

closes: https://github.com/opendatahub-io/odh-dashboard/issues/1370

## Description


Added a new annotation to workbenches called `opendatahub.io/image-display-name` which stores the display name of the selected image. this gets used when displaying the image on the notebook table if the image itself could not be found. if it is not found it will also include the wording `(deprecated)`.

When you got to edit this workbench with a deprecated image. the image will be _unselected_ and require you to select a new one

<img width="547" alt="image" src="https://github.com/opendatahub-io/odh-dashboard/assets/12587674/fe28f22b-26e2-421e-a4a2-d4b15f505028">
<img width="421" alt="image" src="https://github.com/opendatahub-io/odh-dashboard/assets/12587674/a1c394a3-f2de-4bc0-9c2c-ebb1f63aa14c">
<img width="423" alt="image" src="https://github.com/opendatahub-io/odh-dashboard/assets/12587674/dbe8125d-7a8b-42b2-aad9-9e0bd453ba2d">


## How Has This Been Tested?




1. create a custom notebook image
2. create a workbench with that custom image
3. delete the custom image
4. verify that deprecated wording is on the workbench table
5. verify that on edit, the image selection is required
6. remove the annotation (`opendatahub.io/image-display-name`) from the notebook resource, and notice that the image becomes `unknown` on the table row

## Test Impact


I added tests to check the use cases of unknown and deprecated. In doing this i felt the need to implement a few helpers for mocks. namely a `opts: RecursivePartial<k8sKindType>` and then lodash _.merge the base mock with the opts. this gives you the option for fine grained control without having to make a large amount of of options to feed in.

## Request review criteria:



Self checklist (all need to be checked):

If you have UI changes: 


After the PR is posted & before it merges:
